xmtp_mls/src/groups/validated_commit_v2.rs

```diff
@@ -0,0 +1,465 @@
+    // The actual installations added and removed via proposals in the commit
+    let mut added_installations: HashSet<Vec<u8>> = HashSet::new();
+    let mut removed_installations: HashSet<Vec<u8>> = HashSet::new();
+    let mut credentials_to_verify: Vec<CommitParticipant> = vec![];
+
+    for proposal in staged_commit.queued_proposals() {
+        match proposal.proposal() {
+            // For update proposals, we need to validate that the credential and installation key
+            // are valid for the inbox_id in the current group membership state
+            Proposal::Update(update_proposal) => {
+                credentials_to_verify.push(update_proposal.leaf_node().try_into()?)
+            }
+            // For Add Proposals, all we need to do is validate that the installation_id is in the expected diff
+            Proposal::Add(add_proposal) => {
+                // We don't need to validate the credential here, since we've already validated it as part of
+                // building the expected installation diff
+                let leaf_node = add_proposal.key_package().leaf_node();
+                let installation_id = leaf_node.signature_key().as_slice().to_vec();
+                added_installations.insert(installation_id);
+            }
+            // For Remove Proposals, all we need to do is validate that the installation_id is in the expected diff
+            Proposal::Remove(remove_proposal) => {
+                let leaf_node = openmls_group
+                    .member_at(remove_proposal.removed())
+                    .ok_or(CommitValidationError::SubjectDoesNotExist)?;
+                let installation_id = leaf_node.signature_key.to_vec();
+                removed_installations.insert(installation_id);
+            }
+
+            _ => continue,
+        }
+    }
+
+    Ok((
+        added_installations,
+        removed_installations,
+        credentials_to_verify,
+    ))
+}
+
+/// Generates an expected diff of installations added and removed based on the difference between the current
+/// [`GroupMembership`] and the [`GroupMembership`] found in the [`StagedCommit`].
+/// This requires loading the Inbox state from the network.
+/// Satisfies Rule 2
+async fn extract_expected_diff<
+    'conn,
+    'client,
+    'diff,
+    ApiClient: XmtpMlsClient + XmtpIdentityClient,
+>(
+    conn: &'conn DbConnection<'conn>,
+    client: &'client Client<ApiClient>,
+    existing_group_context: &GroupContext,
+    new_group_context: &GroupContext,
+) -> Result<(GroupMembership, InstallationDiff, Vec<Inbox>, Vec<Inbox>), CommitValidationError> {
+    let old_group_membership = extract_group_membership(existing_group_context)?;
+    let new_group_membership = extract_group_membership(new_group_context)?;
+    let membership_diff = old_group_membership.diff(&new_group_membership);
+    let added_inboxes = membership_diff
+        .added_inboxes
+        .iter()
+        .map(|inbox_id| Inbox {
```

Comment:
might be worth `impl From<String> for Inbox` and then this becomes `.iter().map(Into::into).collect<Vec<Inbox>>()`

Reply:
In the next PR I add the `is_creator` flag, which requires some more data than just the Inbox ID